### PR TITLE
AOT: Refactor map creation and runtime resource code

### PR DESCRIPTION
**Pull request by danobi**
This PR refactors map creation and runtime resource collection. This
is mostly a non-functional change with a few bug fixes. I think it's also
a fairly nice cleanup for the code base as well.

In broad strokes, this PR:
* Makes the type information stored in AST nodes more complete
* Splits out map metadata collection and map creation code from
  `SemanticAnalyser` and moves it into `ResourceAnalyser` pass
* Splits out async argument collection from `SemanticAnalyser` and
  moves it to `ResourceAnalyser`
* Encapsulates all runtime metadata in `RequiredResources`

This is generally good (in my opinion) because `SemanticAnalyser`
and `BPFtrace` were a bit bloated with responsibilities. By splitting
responsibility out, we can better encapsulate and contain complexity.

This PR is a major building block for AOT b/c we'll later add
serialization/deserialization capabilities to `RuntimeResources`. The
goal is to make a bpftrace script fully portable by transporting around
BPF bytecode and `RuntimeResources`.

I've included more details in each individual commit.

##### Checklist

- [ ] Language changes are updated in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by danobi** [comment]
I've tried to make each commit an easy to review change. If the PR is still too big I can break it up into multiple PRs.

**Review by fbs — COMMENTED** [review]
Looks good :) Glad that new pass infra works well enough, more small passes is a nice improvement.

File: src/ast/resource_analyser.cpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  }
+  else if (builtin.ident == "kstack" || builtin.ident == "ustack")
+  {
+    resources_.stackid_maps.insert(StackType{});
+  }
+}
+
+void ResourceAnalyser::visit(Call &call)
+{
+  if (call.vargs)
```
[the remark this answers is not in the document]
Comment (danobi) [reply src/ast/resource_analyser.cpp:75]: For some reason it works for `Call &call` but not for `Probe &probe`. The latter causes infinite recursion. It'd be great if it always worked but I'm not sure why it doesn't right now.

[the remark this answers is not in the document]
Comment (fbs) [reply src/ast/resource_analyser.cpp:75]: I think you might be calling `Visit` instead of `visit` so you're dispatching wrong

**Review by fbs — APPROVED** [review]

**Pull request merged**